File: Data Structures/Stack and Queue/QueueThroughLinkedLists.js
// Queue implementation through Linked lists
// FIFO

class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}


class Queue {
    constructor() {
        this.first = null;
        this.last = null;
        this.length = 0;
    }

    // Peek on the first node
    peek() {
        return this.first;
    }

    // Add a node to the queue
    enqueue(value) {
        const newNode = new Node(value);
        if (this.isEmpty()) {
            this.first = newNode;
            this.last = newNode;
        }
        else {
           this.last.next = newNode;
           this.last = newNode;
        }
        this.length++;
        return this
    }

    // Remove the first node
    dequeue() {
        if (!this.first){
            return null;
        }
        if (this.first === this.last) {
            this.first = null;
        }
        
        this.first = this.first.next;
        this.length--;
        return this;

    }

    // Check if the queue is empty
    isEmpty() {
        if (!this.first) {
            return true; }
        return false;
        }
}

// Print outs
const myQueue = new Queue();
myQueue.isEmpty();
myQueue.enqueue('Joy');
myQueue.enqueue('Matt');
myQueue.enqueue('Pavel');
myQueue.enqueue('Samir');
myQueue.peek();
myQueue.dequeue();
myQueue.peek();

File: Data Structures/Stack and Queue/QueueThroughLinkedLists.py
# Queue implementation through linked lits
# FIFO

class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class Queue(Node):
    def __init__(self):
        self.first = None
        self.last = None
        self.length = 0
    

    # Peek on the first node
    def peek(self):
        return self.first
    

    # Add a node to the queue
    def enqueue(self,value):
        super().__init__(value)
        new_node = Node(value)
        if self.is_empty():
            self.last = new_node
            self.first = new_node
        else:
            self.last.next = new_node
            self.last = new_node
        self.length += 1
        return self.value


    # Remove the first node
    def dequeue(self):
        if not self.last:
            return None
        elif self.first == self.last:
            self.first = None
        else:
            self.first = self.first.next

        self.length-=1
        return self

    
    # Check if the queue is empty
    def is_empty(self):
        if not self.last:
            return True
        return False

    

# Print outs
my_queue = Queue()
print(my_queue.is_empty())
my_queue.enqueue('Joy')
my_queue.enqueue('Matt')
my_queue.enqueue('Pavel')
my_queue.enqueue('Samir')
print(my_queue.peek())
my_queue.dequeue()
print(my_queue.peek())
print(my_queue.is_empty())
my_queue.dequeue()
my_queue.dequeue()
my_queue.dequeue()
print(my_queue.peek())
